lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -6498,8 +6498,56 @@ ModulePass *llvm::createSROA_Parameter_HLSL() {
 //===----------------------------------------------------------------------===//
 
 namespace {
+
+struct GVDebugInfoPatchCache {
+  std::unordered_map<Function *, SetVector<DISubprogram *>>
+      SubprogramsForFunction;
+  std::unordered_set<DILocation *> Seen;
```

Comment:
It would likely be much more efficient to use `llvm::DenseMap` and `llvm::DenseSet` here. The allocation overhead of the std types is pretty steep which is why the [LLVM Programmers Manual recommends against using the `std` types](https://llvm.org/docs/ProgrammersManual.html#other-map-like-container-options).